### PR TITLE
Refactored RC control for GPIO pins 

### PR DESCRIPTION
**Pull request by zlite**
Moved RC control to the Complete template and made it so you can use both RC in and out on the GPIO pins at the same time. This will support a RC hat for the Rpi.

Other changes:

--Got rid of the "USE_RC" config setting and instead conformed to standard for other controllers and actuators by creating a "pigpio_pwm" type for both
--I've done a global replace of the "SERVO_ESC" drivetrain type to "I2C_SERVO", which says what it is properly.
--Upped default loop speed to 40Hz

Tested with RC and joystick input, RC and I2C servo board output

Not working yet: auto record on throttle when in RC mode. Will refactor to work like Robohat MM1 part to gain that functionality

### CODE REVIEW

**Comment by zlite** [comment]
Update: I'm refactoring to follow the model of the Robohat MM1 class, which will allow it to work like the rest of the controller parts.

**Comment by DocGarbanzo** [comment]
I've converted your PR to draft. Just click on 'Ready for review' when you are done with all changes.

**Comment by zlite** [comment]
This works in testing.

**Comment by zlite** [comment]
@DocGarbanzo What do you think the Channel 3 function should default to? Recording on/off? Mode switch (user/pilot)? Keep in mind that most people won't have a Ch 3 on their transmitter, so the most typical use case will be to control both of those functions via the web app

**Comment by zlite** [comment]
PS: the linter error shown is bogus; the if statement with Joystick is not reached in the instance that it's a joystick class is not invoked. If a joystick class is invoked the import is done conditionally with "from donkeycar.parts.controller import get_js_controller, JoystickController". For some reason the linter is not smart enough to understand conditional imports. If need be, I can make the import unconditional and waste a little memory to make the linter happy ;-)

            if (cfg.CONTROLLER_TYPE != "pigpio_pwm") and (cfg.CONTROLLER_TYPE != "MM1"):  # these controllers don't use the joystick class
>               if isinstance(ctr, JoystickController):
E               UnboundLocalError: local variable 'JoystickController' referenced before assignment

**Comment by DocGarbanzo** [comment]
I have the `TubWiper` action attached to my channel 3, so I can delete bad images immediately. I think was in the `basic.py` template. But I see that recording/on off would also make sense. Maybe you make this switchable by another flag in `cfg_complete.py` so folks can choose what they prefer.

The import error from GH actions is not a Linter error but a runtime error which you should see locally on your machine as well. This needs to get fixed, maybe the test just requires an update to the config?

**Comment by zlite** [comment]
I fixed the Joystick import issue

**Comment by DocGarbanzo** [comment]
@zlite - can you squash all your commits into one?

File: donkeycar/templates/complete.py
Comment (haslinghuis) [inline donkeycar/templates/complete.py:707]: :space_invader: 
```suggestion
    elif (cfg.CONTROLLER_TYPE != "pigpio_rc") and (cfg.CONTROLLER_TYPE != "MM1"):
```

File: donkeycar/parts/controller.py
Comment (haslinghuis) [inline donkeycar/parts/controller.py:344]: Please remove comments or make comment why this should  be here.

File: donkeycar/parts/controller.py
Comment (haslinghuis) [inline donkeycar/parts/controller.py:1687]: I'll leave this to maintainers.

**Review by haslinghuis — COMMENTED** [review]
Just a few styling and other comments

**Comment by zlite** [comment]
@DocGarbanzo I'm afraid after lots of trying, I still can't figure out how to squash my commits into one without rebasing and screwing up all the history on Dev. The "Squash and merge" button here is grayed out for me; is it available to you?

**Comment by zlite** [comment]
@DocGarbanzo This has all been cleaned up and tested. Ready for your review again (or advice on how to squash commits to make it easier for you)

**Comment by haslinghuis** [comment]
From the `donkeycar` project folder you can use somethink like: https://www.scraggo.com/how-to-squash-commits.
```
git rebase -i HEAD~17
```

- You should see a list of commits, each commit starting with the word “pick”.
- Make sure the topmost, first commit says “pick” and change the rest below from “pick” to “squash”. This will squash each commit into the previous commit, which will continue until every commit is squashed into the first commit.
- Save and close the editor.
- It will give you the opportunity to change the commit message. What you see is a single message containing all of the commit messages. Edit these as you wish.
- Save and close the editor again.
- Important: If you’ve already pushed commits to origin, and then squash them locally, you will have to force the push to your branch.
```
git push origin RCcontrolV2 --force
```

**Comment by DocGarbanzo** [comment]
@zlite - have a look at @haslinghuis recipe. I'm sure you can do this in VS code, too. Firstly, re-base your branch on dev. Then  squash all your commits into one. Both action will require a force push. This force push goes to your branch (not to dev). Then the PR has a single commit and all history remains clean. I can help you if you get stuck.

File: donkeycar/parts/actuator.py
Comment (DocGarbanzo) [reply donkeycar/parts/actuator.py:69]: Please remove these commented lines.

File: donkeycar/parts/actuator.py
Comment (DocGarbanzo) [inline donkeycar/parts/actuator.py:256]: The `(float)` looks like an accidental C-style cast. Pls remove. Also, can you please explain where the magic 0.158 and 0.136 are coming from? Maybe these should be some CONSTANTs?

File: donkeycar/parts/actuator.py
Comment (DocGarbanzo) [inline donkeycar/parts/actuator.py:262]: I would suggest to remove any delay here to get data as recent as possible.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:239]: There is only one `pigpio.pi()` object per RPi.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:247]: Instead of having triplication of all variables here, I suggest to use a list of len 3, that make the whole code more readable and shorter, because you'll just have to loop/iterate over these further down.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:262]: Should be list of 3.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:274]: Then this will become a loop.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:278]: Pls remove these obsolete lines.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:307]: This will become a loop here, too.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:319]: Pls remove the argument `img_arr`.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:337]: These four lines are simply: `is_action = abs(signal2 - self.jitter) > 1.0`

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:350]: You are missing to return `signal3` here. I believe you want to use the `is_action` in order to switch the recording on/off by checking if the throttle is pressed. But we should also return signal 3 here, which for a normal RC control would only give you ON/OFF (or 1500 / 2000 in us). Then other parts in the car can use that channel.

File: donkeycar/parts/controller.py
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:322]: Can you update the doc string so it really describes what it is doing now?

File: donkeycar/templates/complete.py
Comment (DocGarbanzo) [inline donkeycar/templates/complete.py:337]: Obsolete blank line.

File: donkeycar/templates/complete.py
Comment (DocGarbanzo) [inline donkeycar/templates/complete.py:339]: Remove commented line.

**Review by DocGarbanzo — CHANGES_REQUESTED** [review]

File: donkeycar/parts/actuator.py
Comment (zlite) [reply donkeycar/parts/actuator.py:69]: Done

File: donkeycar/parts/actuator.py
Comment (zlite) [reply donkeycar/parts/actuator.py]: Already fixed

File: donkeycar/templates/cfg_complete.py
Comment (zlite) [reply donkeycar/templates/cfg_complete.py:24]: Done

File: donkeycar/templates/complete.py
Comment (zlite) [reply donkeycar/templates/complete.py:206]: There is a class of controllers that all need a different V.add that includes user/mode. That's the reason for the nested ifs. If you really don't like it, I can refactor as you suggest but it just means that I'll have to add a V.add for each of them rather than batching them as I've done here

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:335]: Yes, but in a subsequent commit (sorry again for not squashing!) I added another condition to avoid turning on recording when not wanted. Now that it's a bit more complex, I think it's okay to spell it out a bit more 
` if (abs(signal2 - self.jitter) > 1.0) and self.RECORD: # is throttle above jitter level? If so, turn on auto-record `

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:350]: Done

File: donkeycar/templates/complete.py
Comment (zlite) [reply donkeycar/templates/complete.py:337]: Done

File: donkeycar/templates/complete.py
Comment (zlite) [reply donkeycar/templates/complete.py:339]: Done

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:320]: To be honest, I don't know that that string does but I apparently need it or I get an error :-(

It's harmless

File: donkeycar/parts/actuator.py
Comment (zlite) [reply donkeycar/parts/actuator.py:256]: This is all legacy stuff that I didn't write. No idea what those constants are, but I can at least remove the weird float cast

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:239]: It works as is, so I'm confused by this

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:247]: This seems like just a style preference thing.  looping over lists just isn't my preferred way to do it for only three variables. I find it harder to read.

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:259]: As per above

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:271]: As per above

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py]: I turned these into a Debug option

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:317]: It breaks if I do, for reasons I can't be bothered to figure out. It's harmless as is

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py:305]: As per above

File: donkeycar/parts/actuator.py
Comment (zlite) [reply donkeycar/parts/actuator.py:262]: Done

File: donkeycar/parts/controller.py
Comment (zlite) [reply donkeycar/parts/controller.py]: Actually uncommented. They're need for a test

**Comment by zlite** [comment]
Okay, I realized the squashing problem was that I had based my branch off of upstream dev and has getting everyone else's commits. I rebased on my own repo and was now able to squash just my own commits. 

I think it should all be ready to review again

File: donkeycar/parts/actuator.py
Comment (DocGarbanzo) [reply donkeycar/parts/actuator.py]: Ok, yeah I saw later that you just moved the code.

**Review by DocGarbanzo — APPROVED** [review]
LGTM.

**Pull request merged**
